### PR TITLE
Consistently use GCP XML API

### LINKED ISSUE #4209: Use XML API for GCS
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

GoogleCloudStorage currently uses a mix of the XML and JSON APIs. This is unfortunate and causes issues due to some functionality not being properly supported in the JSON APIs.

One challenge is fake-gcs-server lacks proper support for the XML APIs - https://github.com/fsouza/fake-gcs-server/issues/331

**Describe the solution you'd like**

Switch to using the XML API for GCS

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Relates to #2241
Relates to #3027
Closes #4209

# Rationale for this change
 

We currently use a mixture of the XML and JSON APIs. 

The XML API is more complete, supporting the full set of request preconditions, list offsets, etc... and so we should just switch over. As an added bonus this allows for code sharing between the implementations.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
There appear to be some emulator quirks... Will fix tomorrow...

File: object_store/src/aws/client.rs [hunk 1]
```diff
@@ -118,69 +118,6 @@ impl From<Error> for crate::Error {
     }
 }
 
-#[derive(Debug, Deserialize)]
```
Comment (tustvold) [inline object_store/src/aws/client.rs:121]: This is just moved into list.rs so it can be shared with the gcs implementation

**Comment by tustvold** [comment]
It would appear that fake-gcs-server doesn't properly support the XML APIs... - https://github.com/fsouza/fake-gcs-server/issues/331

File: .github/workflows/object_store.yml [hunk 2]
```diff
@@ -96,7 +96,7 @@ jobs:
 
       - name: Configure Fake GCS Server (GCP emulation)
         run: |
-          docker run -d -p 4443:4443 fsouza/fake-gcs-server -scheme http
+          docker run -d -p 4443:4443 tustvold/fake-gcs-server -scheme http -backend memory -public-host localhost:4443
```
Comment (tustvold) [inline .github/workflows/object_store.yml:99]: https://github.com/fsouza/fake-gcs-server/pull/1164

File: object_store/src/gcp/mod.rs [hunk 3]
```diff
@@ -639,12 +614,6 @@ impl ObjectStore for GoogleCloudStorage {
     }
 
     async fn get_opts(&self, location: &Path, options: GetOptions) -> Result<GetResult> {
-        if options.if_modified_since.is_some() || options.if_unmodified_since.is_some() {
```
Comment (tustvold) [inline object_store/src/gcp/mod.rs:642]: We can now support this :tada:

**Review by alamb — APPROVED** [review]
Looks like a nice cleanup to me -- I read the PR and it all made sense to me.

I am not an expert in the various object store APIs so I probably wouldn't catch some subtle API misuse, but all in all it looks like a nice improvement to me

Thank you @tustvold

**Pull request merged**